### ISSUE TITLE
Geotagging - identify relevant tweets

### ISSUE THREAD

**Issue by alinapark** [1712n/challenge#78]
The goal of this challenge is to create a model to identify tweets referring to the user's location.

### Deliverable:
The goal of the challenge is to create a model to flag tweets containing information useful for identifying the user's location. The model should consume texts on the input, and return a confidence score for the tweet to be relevant to identifying the user's location. The applicant should also provide the threshold confidence value for the interpretation of the score.

For example: "born and raised here! Yankees forever!" would have a higher confidence score than "AWS has been down for a while now" or "my favorite cities are Paris, Milan and London"

The relevancy of a tweet should be evaluated by predicting the author's location, and comparing the prediction with an actual location (as per the training data values)

### Training Data:
The training data set may be found [here](https://drive.google.com/file/d/1tamB8N6RW_K0yVMR1yNPUvppivuGe_tG/view?usp=sharing). It includes over 8M tweets, along with the user's geographical coordinates (latitude and longitude) and basic metadata, such as language and the time of post. 

#### Results Evaluation:
We'll evaluate the model with the data set not shared in the challenge. 

To make the evaluation process easier, please make sure your code includes a function that:
- takes a file (of the same format) on the input
- returns accuracy, recall, precision and F1 scores on the output  
- the model's weights are included 

#### To submit your solution:
- message us at challenge-submission@blockshop.org with a link to this issue, an overview of your model architecture, the obtained evaluation metrics, and your resume.
- we will contact all applicants with working models, evaluate the model on a held-out test dataset, and schedule a 30-min interview to discuss further steps.
- don't hesitate to ask us questions by commenting in this issue or emailing us at challenge-program@blockshop.org.

---

### Successful submissions

🎉 @orzhan successfully solved the challenge and was hired by @inca-digital

**Comment by hivaze** [comment]
Correct the description of the task, please, it's not clear what needs to be predicted, what does the topic and geolocation have to do with it?

**Comment by alinapark** [comment]
@hivaze the model should analyze a piece of text and predict whether it is useful to identifying the user's location. I added a couple of examples - hope it helps

**Comment by AbdullahMakhdoom** [comment]
Hi @alinapark,
The training dataset does not have a label column, using which the model could train and tell whether a tweet is useful for identifying a user's location or not. For example:  "born and raised here! Yankees forever!" will have a positive label while "my favorite cities are Paris, Milan and London" will have a negative label.

**Comment by alinapark** [comment]
@AbdullahMakhdoom you're right - it doesn't. The task itself is to create a model that will identify geolocation-related texts based on the data provided. The methodology is up to the candidates - whether that's unsupervised learning, annotations, or any other methods

**Comment by joebutcher** [comment]
@alinapark do you have a preference on what level of location we should be predicting? For example, for tweets in the U.S., should we be predicting city, state, closest large metro area, etc.? Any guidance here would be helpful.

**Comment by delattre1** [comment]
@alinapark Should the model work only with texts in English (en), or all those other languages inside the example dataset?

**Comment by evgenydmitriev** [comment]
> @alinapark do you have a preference on what level of location we should be predicting? For example, for tweets in the U.S., should we be predicting city, state, closest large metro area, etc.? Any guidance here would be helpful.

The model performance will be measured based on the distance between predicted and actual coordinates.

> @alinapark Should the model work only with texts in English (en), or all those other languages inside the example dataset?

The model performance will be measured using our test dataset, withheld from the challenge. Just like the training dataset, it contains many non-English tweets.

**Comment by lidiaToropova** [comment]
@evgenydmitriev According to task description 

> The goal of the challenge is to create a model to flag tweets containing information useful for identifying the user's location. The model should consume texts on the input, and return a confidence score for the tweet to be relevant to identifying the user's location.
it's binary classification problem.
But according to your comment 

> The model performance will be measured based on the distance between predicted and actual coordinates.
the model should predict coordinates.

So should the model output a confidence score for the tweet to be relevant to identifying the user's location or predict coordinates?

**Comment by evgenydmitriev** [comment]
@lidiaToropova correct, the only one way you would be able to judge the relevancy of the tweets is by the distance between the predicted and the actual coordinates. The relevancy part is to highlight the importance of high-confidence decisions and to allow candidates to filter out the noise tweets early on.

@alinapark you might want to include and example of a successful submission and expand the evaluation section to make things more obvious.

**Comment by ScientificCollaboration** [comment]
Provided BERT based solution.

**Comment by evgenydmitriev** [comment]
🎉 @orzhan successfully solved the challenge and was hired by @IncaDigital! Congrats 🎉

**Comment by r-log** [comment]
@ScientificCollaboration well I looked up the dataset and everything looks fine from this end. Try to open a dataset with editor when you clean it and see if the coordinates are there for that specific tweet, so that you can be sure that you are cleaning it correctly. If everything matches up after the cleaning then troubleshoot along. In the original un-cleaned dataset provided if you search for the text you can actually see that two tweets are in one row because they aren't separated correctly. You can see the difference if you open it with any csv editor as they tend to merge text when they are not separated with comma correctly or if there are columns with NaN values.
When you open un-cleaned dataset with any regular text editor you can see that the line that you've screenshot doesn't even have a second coordinate it only has one therefore its messing up your dataframe and not showing correctly.

**Comment by ScientificCollaboration** [comment]
ok what are the ground truth data for this tweet? I am not talking about data cleaning in the last  columns. I generated them myself by means on NER extravtion and grouping - that is quite evident. And NaNs there because I had not yet finished processing. That are NLP generated columns. Idea was to create 3 levels of geotagging - country, city,building. They are not finshed and I know that. 
I am talking about so called "ground truth data". Rows n-1 and n+1 are correct but row with tweet from Puco Chile has longitude and attitude 0 and 0. That is impossible and can be verified just by analysing the Globe. Pucon Chile has  Latitude: -39.2822 Longitude: -71.9543. And there are a lot of same errors.  If you have correct data, what are the coordinates of this tweet for example?

**Comment by r-log** [comment]
@ScientificCollaboration There is no longitude data for that tweet it is evident when you open un-cleaned dataset in any regular text editor and search for the tweet provided in a screenshot, you can see that the tweet doesnt have second coordinate it only has latitude.. that means you didnt cleand data properly cause you need both of values for everything else to work. Just open the dataset provided from here and search for tweet you provided in screenshot. You will see that row for specificly that tweet doesnt have longitude it only has lat. And I think thats the reason it messes up with everything else.

**Comment by ScientificCollaboration** [comment]
I see both lon and lat in any case, but there are a lot of missed data in lon and lat. I forgot that I substituted nan with zeros instead of removing rows completely. But in any case data is strange - even with missed timestamps(?). Almost 30 % of bad values that is too much. https://github.com/ScientificCollaboration/Geocodes/blob/main/error2.jpg
May be there was another post where someone definetly exposed their location ("like we are here near the Effifel Tower") but the location was  different.

**Comment by r-log** [comment]
@ScientificCollaboration yes exactly that is the case. You need to first get rid off all the rows with missing column values cause they are non relevant to us in this case. But you need to do it in a smart way because as I told before columns are separated with commas that's why csv is called csv, it is comma separated values (csv).The row must be complete and have all the values for each column, if one value is missing from the column in one row then majority of data and rows after that will be shown wrong.

example: columns are tweet_text,lang,timestamp,lat,lon
so if you have a row that has each value in these columns they are good 
example: "Nak berak masa berlari is another level of annoying🥴",in,2021-08-24T00:20:01.000Z,**3.14853065**,**101.705305**

If we look at your tweet: "Acaba de publicar una foto en Pucón,Chile",es,2021-08-11T02:54:15.000Z,**-39.29045965**
it misses the last value so it continues and merges next tweet with this one on tweet_text column and puts it on that same row and creates invalid data. To prevent that for happening you should put comma at the end of the row where column value is missing

example: "Acaba de publicar una foto en Pucón,Chile",es,2021-08-11T02:54:15.000Z,-39.29045965,<-------
in this case when you put comma at the end it will read it as an empty column and it wont scramble next data in row.

**Comment by ScientificCollaboration** [comment]
I dont't have these tweet - ""Acaba de publicar una foto en Pucón,Chile",es,2021-08-11T02:54:15.000Z,-39.29045965"? Where is it? I watched all the panda frames. All they have NAN both lon and lat.

**Comment by r-log** [comment]
@ScientificCollaboration its in the dataset they provided us with, [here](https://drive.google.com/file/d/1tamB8N6RW_K0yVMR1yNPUvppivuGe_tG/view) row: 6251

**Comment by ScientificCollaboration** [comment]
Still don't see it there, but there are a lot of other incomplete data of other kind. I solved this issue when I calculated Vincenty  distances because incomplete coordinates are not processed at all. I removed simply all shifted and uncomplete data but there should be even more data to be removed.

**Comment by r-log** [comment]
There is a lot of incomplete data indeed, that's why this is a challenge. But that's why you need to firstly **CLEAN YOUR DATA!**, because the dataset is messed up. 

There are characters in dataset like --->" and extra commas that are creating problems with reading the data into a data frame. You can't feed the model with dataset that is messed up because later on eventually some of these errors will interfere with results of a prediction. 

Just take your time download .csv editor (because pandas show only data that it can be read!!) pandas will not show **BAD LINES** if you read it into dataframe. You cannot even read this dataset normally like this --->**pd.read_csv("filename.csv")**  you need to write it like --->**pd.read_csv("filename.csv", on_bad_lines="skip")** to even read it into a dataframe, but if you skip bad lines then less then half of your data will not be shown and there will still be overlapping columns because of the separators.

Please do a research on how csv files work and how pandas is reading the csv you inputed, when you have the basics covered then examine the given dataset both in **CSV EDITOR** as well as **NOTPAD** find lines that are creating a problem  **CLEAN IT!** and then you can feed it to your model.

**Comment by ScientificCollaboration** [comment]
Notepad? Is that a joke? I would rather write my OWN parser checking completenes of data in each line. That would be much more logical.  But that is outside the scope of average DS task.  Listen I did that long ago axactly with pd.read_csv("filename.csv", on_bad_lines="skip" .... other features) . And processed the data - typecasting, NAN removal and other. I don't want to undertand what other mistakes were made in that data. Because there  can be thousand other mistakes, shifts, data loses, misses etc.
The problem it that this is completely unacceptable data with very poor quality. Cleaning data is not correcting mistakes created by bad work of Data Engineers. Cleaning -  it is imporving statistical data - missed values, noise (only if it is acceptable), outliers,gradients, collinearity, heteroskedasdisity, gaussian noise etc.  Here data is completely corrupted - some parts are shifted some swallowed somewhere. There must be algorithm parsing data and finding wheter it is complete line or not. No any guaranty that proper line will be recovered even after this corrections, because text in tweet can contain labels similar to standard endings or the beginnings - tweet ID or coordinates. And this is Data Engineering. 
Here
 "
4295,"Woohoo!  The new patio is done!  Next up, California https://t.co/sZIfTv2Tf9",United States,en,2021-08-13T01:38:17.000Z,
4296,Beloved of the LORD in @Bahrain #TheGlorifiedBlood of JESUS CHRIST can wash away our sins and transform our sin into righteousness and transfigure,,,,
4297,"🤸🤸
#repostando em Barra do Gramame https://t.co/tmstW71L8v",pt,2021-08-15T23:42:42.000Z,-7.1494765000000005,-34.88337085
4298,Why don’t you go to India https://t.co/9fYBkXNKBr,en,2021-08-04T20:09:47.000Z,52.4939665,-1.8906405
4299,"Damm Grandma 😜😅😅😊" - No real understanding why it shows coords and why not or here 
"4304,"Que vergonha esse time do Brasil,Brésil,pt,2021-08-05T05:43:06.000Z,-22.9114217
¡Información importante!",,,,
4305,Hoy miércoles 18 de agosto y mañana jueves 19 de agosto,,,,
4306,"@santialvarezl @Danvelasq4 @AlcaldEstrella @LiliMariaR""",es,2021-08-18T13:40:13.000Z,6.144337997625265,-75.6639979962697
4307,"I'm at the Gillands in Lancaster,United States,en,2021-08-29T15:48:16.000Z,33.6311395
At 10:05 PM EDT", NC] FIRE DEPT/RESCUE reports LIGHTNING. LIGHTNING STRIKE TO ROOF WITH TORRENTIAL RAIN," FULL ROOF COLLAPSE WITH BUILDING ON FIRE https://t.co/0frcB5uoc2""",United States,en"
This is not task to clean the data, this is task to read data from DB properly.

**Comment by sultanovazamat** [comment]
Hi! sent you the solution, could you please check out the inbox? Thanks!

**Comment by sultanovazamat** [comment]
@evgenydmitriev @alinapark How long does it usually take for you to check the provided solution? It's been more than 2 weeks since I sent it:(
